### DISCUSSION TITLE
Saving normalized array to PNG

### DISCUSSION THREAD

**Discussion by WerctFourth** [libvips/libvips#4022] [Q&A] [answered]
Hello.
I use pyvips with numpy. I load an image, convert it into a normalized array (0.0-1.0 values) and process it in my script. Then I convert the array into a pyvips image (using normalized values without any conversion) and save it.
It works fine with RGB images, libvips interprets the array as scRGB and saves it. But when I try to save a grayscale image, it's interpreted as b-w double and the saved image is just a black 8-bit PNG.
Is there a way to save a single band image from a normalized array in libvips?

**Comment by jcupitt** [comment, marked as answer]
Hello @WerctFourth,

No, libvips only has `scrgb`, it doesn't have a tag for 0-1 greyscale images.

You could save as TIFF -- it supports float pixels:

```
$ vips colourspace k2.jpg x.v scrgb
$ vips extract_band x.v x2.v 1
$ vipsheader x2.v
x2.v: 1450x2048 float, 1 band, scrgb, jpegload
$ vips max x2.v
1.000000
$ vips copy x2.v x.tif
$ vipsheader x.tif
x.tif: 1450x2048 float, 1 band, b-w, tiffload
$ vips max x.tif
1.000000
```

Formats like PFM, nifti, MAT and FITS would work too.

**Reply by WerctFourth** [reply]
Thank you for your answer.
That's unfortunate that it can't be done directly.
In this case, I've found two hack-ish solutions for my situation:
1. Cast this array to uint16 manually and save it as a 16-bit grayscale PNG;
2. Just multiply this single band into three RGB channels and save it as an RGB PNG.

**Reply by jcupitt** [reply]
OK, that sounds reasonable.
